Correctly paint the CSS canvas’ background

SimonSapin opened the pull request:
https://drafts.csswg.org/css-backgrounds/#special-backgrounds
    
Fixes https://github.com/servo/servo/issues/25559
Closes https://github.com/servo/servo/pull/26121, as it is an alternative.
Fixes https://github.com/servo/servo/issues/26444.

mrobinson committed:
Update WPT expectations

SimonSapin committed:
`display: contents` computes to `block` on the root element

SimonSapin committed:
Set in the DOM the box generated by the root element

SimonSapin committed:
Rename BoxTreeRoot/FragmentTreeRoot to BoxTree/FragmentTree

SimonSapin committed:
Correctly paint the CSS canvas’ background

https://drafts.csswg.org/css-backgrounds/#special-backgrounds

Fixes https://github.com/servo/servo/issues/25559
Closes https://github.com/servo/servo/pull/26121, as it is an alternative.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/node.rs
 * @jgraham: tests/wpt/reftests-report/gen.py, tests/wpt/reftests-report/prism.css, tests/wpt/reftests-report/prism.js
 * @kichjang: components/script/dom/node.rs
 * @emilio: components/style/values/specified/box.rs

SimonSapin commented:
At the moment this PR has the same test expectation changes as https://github.com/servo/servo/pull/26121. It passes a couple more tests, but has some remaining regressions on my machine.

@bors-servo try=wpt-2020

* These two tests dynamically set `display: none` to the root element. They hang without output, I haven’t investigated them yet.
  ```
  ▶ TIMEOUT [expected PASS] /css/css-backgrounds/background-color-root-propagation-002.html
  ▶ TIMEOUT [expected PASS] /css/css-backgrounds/background-color-body-propagation-006.html
  ```

* A transform on the root element is not applied to the positioning area of the canvas’s background images. @mrobinson Would calling `build_reference_frame_if_necessary` on the root `StackingContext` be appropriate here?

  Also, the *painting area* is a rectangle in the top-level (un-transformed) coordinate space, but the display item’s `bounds` rect is in local (transformed) coordinates. So we may need to take the axis-aligned bounding box of the inverse-transformed rect.

  ```
  ▶ FAIL [expected PASS] /css/css-transforms/transform-translate-background-002.html
  ▶ FAIL [expected PASS] /css/css-transforms/transform-translate-background-001.html
  ```

bors-servo commented:
:hourglass: Trying commit 70f2eadcfdf44f5e6a48fa4b959fa324e6e97890 with merge 5454e9bef1a0645e9ede290acd4ef2bb34929c70...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/VrOrFo-8QW-v96_bK0H4yQ)

mrobinson commented:
> A transform on the root element is not applied to the positioning area of the canvas’s background images. @mrobinson Would calling build_reference_frame_if_necessary on the root StackingContext be appropriate here?

This depends: Is the other content on the page transformed? If it is, the correct solution would likely be to assign the background display item the correct reference frame as its spatial node. If the rest of the page content is not transformed (but should be), then I think appropriate thing would be to create a new reference frame at the the top level.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26407) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin committed:
Rewrite handling of pointers to boxes in the DOM tree

SimonSapin commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 0a39c6ab58e0785a24d707cedc9ac8bb6b9fc494 with merge ce9c20f89f9ad464a0e42112c1dbdbe196956a74...

SimonSapin commented on components/layout_2020/display_list/stacking_context.rs:
```diff
@@ -244,7 +244,107 @@ impl StackingContext {
+        //   (in local coordinates, which may be rotated from top-level coordinates)
+        //   bounding box of those two transformed rectangles.
+        //
+        // Instead, let’s use a huge rectangle that will hopefully contain everything.
+        //
+        // `f32` has 24 bits of mantissa.
+        // 20 bits of magnitude leaves a few bits of precision
+        // for `background-position` computation to be accurate
+        // to more than a device pixel on very high DPI.
+        const BIG: f32 = (1u32 << 20) as f32;
+        let painting_area = &LayoutRect::new(
+            LayoutPoint::new(-BIG / 2., -BIG / 2.),
+            LayoutSize::new(BIG, BIG),
+        );
```
```
  ▶ CRASH [expected FAIL] /_mozilla/css/mix_blend_mode_a.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  │ [2020-05-07T12:17:09Z ERROR webrender::render_task] Attempting to create a render task of size 1048576x1048576
  │ explicit panic (thread WRRenderBackend#0, at /root/.cargo/git/checkouts/webrender-c3596abe1cf4f320/4ba17a6/webrender/src/render_task.rs:37)
```

Ok so using a huge rectangle might not be a good idea if WebRender wants to actually allocate that many pixels.

SimonSapin replied:
Minimized reproduction:

```html
<html style="background: white">
<body style="mix-blend-mode: difference">
```

SimonSapin replied:
The same panic happens for different inputs on `master`. I’ve filed https://github.com/servo/servo/issues/26463.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26477) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@mrobinson Do you want to review Simon's change here?

mrobinson commented on components/layout_2020/display_list/stacking_context.rs:
```diff
@@ -244,7 +244,107 @@ impl StackingContext {
+        macro_rules! debug_panic {
+            ($msg: expr) => {
+                if cfg!(debug_assertions) {
+                    panic!($msg)
```
Maybe emit a warning for release mode so that it's obvious something is wrong?

SimonSapin replied:
CI runs WPT with debug assertions. Do we expect people who do manual testing to look at a warning log?

mrobinson replied:
Emitting a warning in release mode might be interesting if, for instance, this bug became quite common but only for pages on the web. If you feel like this isn't useful though, feel free to ignore this.

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -38,6 +40,10 @@ pub struct BoxTree {
     root: BlockFormattingContext,
+
+    /// https://drafts.csswg.org/css-backgrounds/#special-backgrounds
+    #[serde(skip)]
```
Is there some reason to not simply implement `Serialize` and `Deserialize` on `CanvasBackground`?

SimonSapin replied:
`ComputedValues` doesn’t implement `Serialize`, but I can push the `skip` attribute down to that field of `CanvasBackground`.

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -53,10 +59,14 @@ pub struct FragmentTree {
     root_fragments: Vec<ArcRefCell<Fragment>>,
 
     /// The scrollable overflow of the root of the fragment tree.
```
This comment might need a small update to reflect the fact that this is for all of the root fragments.

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -359,3 +375,72 @@ impl FragmentTree {
+    /// The element whose style we have above
+    pub from_element: OpaqueNode,
+
+    pub root_element: OpaqueNode,
```
A doc comment might be useful here explaining the distinction between this element and `from_element`.

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -359,3 +375,72 @@ impl FragmentTree {
+    }
+}
+
+fn iter_child_nodes<'dom, Node>(parent: Node) -> impl Iterator<Item = Node>
```
Is this duplicated from `components/layout_2020/dom_traversal.rs` on purpose?

SimonSapin replied:
No, I meant to move it there and forget to remove the original :)

mrobinson commented on components/script/dom/node.rs:
```diff
@@ -3439,12 +3439,18 @@ impl Into<LayoutElementType> for ElementTypeId {
     #[inline(always)]
     fn into(self) -> LayoutElementType {
```
Nice.

mrobinson commented:
Sure. I'm happy to review this. I took a look at this and made a few comments. In general, this all looks good to me. I seem to have found one case where the rendering may not be what is expected. The issue with the infinite rect approach is that this is not always enough to cover the entire canvas. Here's an example:

```html
<style>
html {
    transform: rotateY(90deg);
    background: blue;
}

div {
    transform: rotateY(90deg);
    width: 100px;
    height: 100px;
    background-color: pink;
}
</style>

<html>
    <div>Test</div>
</html>
```

This renders completely blue in Firefox, but no background is rendered with your change (I think).

mrobinson commented:
I should mention that this case is pretty weird. It seems totally reasonable to land this change and gradually cover all of the unusual edge cases (r=me).

SimonSapin committed:
Separate computing a WR reference frame from adding it to the display list

SimonSapin commented:
> This renders completely blue in Firefox, but no background is rendered with your change (I think).

I’ve changed `background-color` to be painted outside of any transform, which should fix this specific case. That leaves `background-image` though.

I’ve also done an attempt at the "proper" inverse transform of the painting area rectangle instead of an arbitrary huge rect, which means we shouldn’t hit https://github.com/servo/servo/issues/26463 unless the document is actually huge. However in doing so I seem to have broken transforms (many WPT failures) so that still needs investigating.

mrobinson commented on components/layout_2020/display_list/stacking_context.rs:
```diff
@@ -464,22 +470,24 @@ impl BoxFragment {
-            new_containing_block_info.rect.origin =
-                (-relative_border_rect.origin.to_vector()).to_point();
+
+        if let Some(reference_frame_data) = &reference_frame_data {
```
Perhaps this could all be part of a method called `build_reference_frame_if_necessary(containing_block_info, new_containing_block_info)`. It might be good to isolate all of the complications of setting up the ReferenceFrame into one method.

SimonSapin committed:
Fix incorrect sign background-image rectangle computation

SimonSapin commented:
> However in doing so I seem to have broken transforms (many WPT failures) so that still needs investigating.

Alright, you gave me the correct lead on Matrix to fix this bug. This uncovered another pre-existing bug, for which I added a separate commit. And I also introduced another bug in passing, now fixed as well. I think this should be in pretty good shape now.

@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit a024ecccba76027bd6d022987b3b96de381c0eaf with merge 13011eecf71fe94b5b4dbad7761b8e8f6d6f2f14...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Ww9cyU-eShm-S95Eh6v3jA)
State: approved= try=True

mrobinson reviewed (approved):
Thanks! These changes look good.

SimonSapin commented:
@bors-servo r=mrobinson

bors-servo commented:
:pushpin: Commit a024ecc has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson a024ecccba76027bd6d022987b3b96de381c0eaf -->

bors-servo commented:
:hourglass: Testing commit a024ecccba76027bd6d022987b3b96de381c0eaf with merge 7d7e987ba052c83c29bc132f7ef5274a008eab31...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/LpBti5TTSzm85eZfjC8DWg)
Approved by: mrobinson
Pushing 7d7e987ba052c83c29bc132f7ef5274a008eab31 to master...
